dace/codegen/targets/unroller.py

```diff
@@ -8,7 +10,137 @@
 from dace.codegen.targets.cpp import sym2cpp
 from dace.codegen import cppunparse
 from itertools import product
+from dace.sdfg import state
+import dace.subsets
+import dace.sdfg
+from dace.sdfg.replace import deepreplace
+from dace.sdfg import nodes as nd
+from dace import Config
 
+def backup_statescope_fields(
```

Comment:
Could you explain why we need these backup functions? We can we not just deep copy, change the deep-copied graph, then pass this to the dispatcher?

Reply:
The reason for this is that deepcopy doesn't work well with subscopes of the graph. There are functions and parts of the codegen, that access state data on the top-level (outside of the copied scope, e.g. on state or sdfg-level) which means you can still get the old objects for some type of calls. This then leads to all sorts of weird behavior.

Reply:
Ok, I understand. As usual, I'm not super comfortable with this, but also don't have an amazing alternative idea.

The only idea I have is whether we could limit the "backup" to be part of the replacement traversal only: When `replace` traverses the SDFG and finds an occurrences of the token that we are replacing, save a backup before replacing it, and return a tuple of replaces entities (similar to what you are doing now). Then, a separate "unreplace" ("replace inversion"? "rereplace"?) understands this format and reverts these changes. The advantage over what you are already doing is that we _only backup what we change_, so we don't risk backing up/restoring too much or too little.

Does this make any sense?

Reply:
Another suggestion: I'm not so sure if it is even so clever to change the actual sdfg during unrolling. In principle, we could do the same thing as with the constants in sdfgs, i.ee we create a local scope for each unrolled section (I believe thats already done in face), and then define a constexpr before each section? The only bad thing I could see is that something like nested unrolling would not work You could still have nested maps with unrolling, but you may not bind the range of the inner map to the variable defined by the outer map.

Reply:
If this works then yes absolutely, I'm not sure about Intel though, since OpenCL doesn't support `constexpr`. Do you want to try?

Reply:
It's definitely not as simple, I doubt you can even do shadowing of OpenCL constant variables. Honestly you're probably better off using `#define` and `#undef` or something nasty like this.

> how can I find out from the unroller if it is currently generating Intel FPGA code?

The dispatcher calls specific code generators, which then call other functions. Some of these other functions take the code generator currently running as an argument (typically a keyword argument `codegen`), but not necessarily all of them. In this case I don't see any easy way for you to know, since this is implemented as its own code generator, so by definition it's neither Xilinx or Intel.

Honestly, we might be better off holding off on Intel OpenCL support. If you can make it work for Xilinx and CPU, then we can leave this pain for another time. @TizianoDeMatteis what do you think?

Reply:
Constants have always been problematic for Intel OpenCL: you can't do shadowing, and, we currently use`#define` to handle some strange situations. I'm also in favor of the `TODO(later)`

Reply:
alright then I'l just do constexpr, add a TODO and get rid of the replacement/backup stuff